### PR TITLE
std: Fix early overflow in time calculation

### LINKED ISSUE #6867: std.time.Timer overflows on Windows after 30 minutes
I'm not sure if timers are meant to go this long but 30 minutes seems like a short enough time span that it should be supported.  The overflow is in `nativeDurationToNanos`:
```zig
return @divFloor(duration * ns_per_s, self.frequency);
```
This multiplication will overflow when duration is >= `18_446_744_074`.
`self.frequency` is `10_000_000` on my system, so this takes 1845 seconds, or about 30 minutes.  Since the frequency is a perfect factor of `ns_per_s`, an ideal conversion wouldn't overflow until the time passed was actually too large to fit into a 64-bit number, which happens after about 584 years.

### PR DESCRIPTION
**Pull request by LemonBoy**
Closes #6867

### CODE REVIEW

**Comment by gereeter** [comment]
This is *probably* fine for this specific use-case since clocks aren't crazily precise, but `safeMulDiv` as you've written it still has the risk of overflowing unnecessarily. For example, if the clock frequency on Windows was 100/ns (`self.frequency = 100000000000`) and the duration was just under a second (`duration = 99999999999`), `q` would be zero, `r` would be `duration`, and `r * ns_per_s` would overflow 64 bits.

Is calculating with 128 bit integers measurably slower?

**Comment by LemonBoy** [comment]
> Is calculating with 128 bit integers measurably slower?

The `int128` division calls `__udivti3` while the `u64` version is much much smaller.

File: lib/std/time.zig [hunk 1]
```diff
@@ -242,15 +242,24 @@ pub const Timer = struct {
...
-            return @divFloor(duration * self.frequency.numer, self.frequency.denom);
+            return safeMulDiv(duration, self.frequency.numer, self.frequency.denom);
         }
         return duration;
     }
 };
 
+// Calculate (a * b) / c without risk of overflowing too early because of the
+// multiplication.
+fn safeMulDiv(a: u64, b: u64, c: u64) u64 {
```
Comment (data-man) [inline lib/std/time.zig:256]:
```zig
fn safeMulDiv(comptime T: type, a: T, b: T, c: T) T
```
and move to math?

Comment (LemonBoy) [reply lib/std/time.zig:256]: It's not meant to be used for anything but this calculation, the safe part in the name refers to the typical values used here.

**Pull request merged**